Provide option to ignore estimated rewind hole for trackerbar->MIDI mapping

broadwell opened the pull request:
Needs to be rebased against master after PR #14 is merged.

broadwell committed:
Revert previous attempt to fix rewind-shifting logic (will be handled here)

broadwell committed:
Add option to ignore rewind hole when mapping tracker columns to MIDI numbers

broadwell committed:
Logic correction: default to using rewind hole-based correction

broadwell committed:
Fix arithmetic of MIDI number reassignment based on rewind hole position

Previously, there was a bug in the adjustment of each hole's MIDI number that
is done when realigning the tracker columns and MIDI numbers based on an
observed discrepancy between the detected rewind hole's MIDI number and the
MIDI number the rewind hole should have given the roll type. The code shifted
the MIDI numbers in the wrong direction, i.e., it interpreted the sign of the
value of `shifting` incorrectly. This value should be *subtracted* from the
MIDI number of each hole to correct the holes' alignment with the virtual
tracker bar, and conversely it should be *added* to the value of the mapping
from MIDI numbers to tracker bar columns (this part was already being done
correctly).

simonwiles commented:
I've just rebased and force-pushed this branch now that #14 is in, but I think (and perhaps you meant?) that it should go in on top of #13.

(The branch in its previous state is preserved at `rewind-hole-toggle-preserved`, in case it's wanted/needed.)

broadwell commented:
> I've just rebased and force-pushed this branch now that #14 is in, but I think (and perhaps you meant?) that it should go in on top of #13.

Yeah, it's probably better to get #13 in before rebasing and eventually merging this one.

simonwiles commented on src/RollImage.cpp:
```diff
@@ -564,7 +576,7 @@ void RollImage::assignMidiKeyNumbersToHoles(void) {
 		return;
 	}
 
-	int shifting = rewindholemidi - midiKey[bestFitIndex];
+	int shifting = midiKey[bestFitIndex] - rewindholemidi;
```
Before I go ahead and merge this otherwise straightforward PR, would you mind elaborating on this change, for my benefit?  The commit message reads "Refine candidate rewind hole detection; ...", but it's not otherwise mentioned.  Thanks!

broadwell replied:
Wow, well spotted -- you could be a code detective. 😉  I rather disgustedly threw in this fix to reverse a change I had made in PR #12, when I was briefly convinced that the pre-existing code was wrong and had been incorrectly reassigning the tracker-column-to-MIDI-number mappings whenever a discrepancy was noticed between the column in which the suspected rewind hole was found vs. the column in which the rewind hole was expected to be given the initial tracker-column-to-MIDI mapping. The sliding-window approach also introduced in #12 correctly resolves most such discrepancies (which are fairly rare anyway), which further helped to confuse the matter at the time.

Looking at the code now, I still have a hard time convincing myself that the value of `shifting` shouldn't be flipped the other way. The root cause is no doubt just that Craig's code is a bit too baroque here, but I haven't really been able to improve on it. For now, the proof is in the aligning; i.e., in all subsequent tests and uses of this code, the change here produced the correct alignment. I'll stare at the code and scratch my head a bit more tonight in the hopes that enlightenment will come; you're welcome to do the same, although I'll certainly understand if you prefer not to.

broadwell replied:
OK, to isolate and test this code I eventually just found a roll for which `tiff2holes` generates an incorrect tracker-to-MIDI-numbers mapping when no corrective code is run (fb623cd1247), commented out the part of `analyzeMidiKeyMapping()` that uses the sliding-window method from #12 to fix the mapping, and then checked to see if the rewind hole-based adjustment in `assignMidiKeyNumbersToHoles()` (which usually runs after the sliding-window method has been applied) was able to shift the assignments correctly. It was not! It actually made the misalignment worse, suggesting that the shift really is being done in the wrong direction. So probably the rewind hole-based alignment correction code has been broken all along, but its failure was masked when generating MIDI file batches recently due to the addition of the sliding-window method.

So there definitely is a problem with the direction of `shifting` in `assignMidiKeyNumbersToHoles()`, but it seems to be easy to fix.  
This mapping adjustment:
`midiToTrackMapping[i] += shifting;`
actually works fine because it adds `shifting` to the track numbers rather than to the MIDI numbers (so the shift is in the correct direction), but it's irrelevant because the MIDI number values in the output MIDI files and the hole report data aren't derived from `midiToTrackMapping` -- they come directly from the hole objects stored in `trackerArray`, and their MIDI numbers are adjusted via the line
`trackerArray[i][j]->midikey += shifting;`
So in order for the rewind hole-based adjustment to produce the proper MIDI number assignments for each note/hole, the shift just needs to be applied subtractively rather than additively:
`trackerArray[i][j]->midikey -= shifting;`

Let me know what you think about this, and if the fix looks acceptable to you, whether it's OK just to add it as another commit to this PR.

Btw, I do think it's worthwhile to keep both the sliding-window and rewind hole-based correction methods in play. The sliding-window method isn't foolproof (it's somewhat susceptible to phantom columns created by tears at the right and left margins), so the rewind-hole method, when implemented correctly(!), is a useful check on it. There are however a few cases in which really weirdly printed rolls trick the rewind hole adjustment into incorrectly overruling the sliding-window adjustment, which is why this PR exists; for those cases, we can just keep a list of those rolls adjacent to the roll-processing script and enable the command-line switch introduced in this PR when they come up.

simonwiles replied:
Aha!  I had begun to look at this and had started to think I was getting somewhere, but I hadn't come close to arriving at this.  I think you take the code detective crown :) 👑

Your explanation and proposed fix make good sense to me -- have you tested it beyond _Si oiseau j'étais_?  Seems like that's a good one to keep a note of for a test suite (automated or otherwise).  **edit:**  I guess I mean: have you confirmed it doesn't change the output for rolls that aren't affected by this and/or are you aware of others where it *is* relevent?

I've no problem with adding a new commit to this PR, but I would suggest a commit message that either recapitulates or references your message above (or comments directly in the code, I suppose -- or both!), since this is not the easiest thing to grok quickly, even with plenty of context.

broadwell replied:
> Your explanation and proposed fix make good sense to me -- have you tested it beyond Si oiseau j'étais? 

Yes, I've tested it on a few more rolls now, including rolls that don't need any rewind hole-based correction, rolls for which rewind hole-based correction should be disabled because it erroneously overrides the sliding window-based result (these were the original motivation for this PR, though they're thankfully rare), and rolls for which the rewind hole-based correction saves the day. They all turned out well. I'm keeping a list of the problematic rolls that could be used in a test suite; so far, my approach has just been to re-process the entire roll corpus periodically as the code evolves and then to spot-check the list of known problematic rolls and a few others to make sure they all seem to play OK, but this is obviously not the most rigorous and reproducible/automated testing method.

broadwell commented:
I've rebased and cleaned up the commit list somewhat, and added some more in-code comments that attempt to explain the ins and outs of the tracker-MIDI alignment correction. I think this should be ready for final review and merging.

simonwiles commented:
This is great, thanks -- sorry it took me so long to get to it.
